src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java

```diff
@@ -207,7 +207,7 @@ private void setUserInfoInThreadContext(User user, Set<String> mappedRoles) {
             StringJoiner joiner = new StringJoiner("|");
             joiner.add(user.getName());
             joiner.add(String.join(",", user.getRoles()));
-            joiner.add(String.join(",", Sets.union(user.getSecurityRoles(), mappedRoles)));
+            joiner.add(String.join(",", Sets.union(user.getSecurityRoles().stream().collect(ImmutableSet.toImmutableSet()), mappedRoles)));
```

Comment:
What does this change do?

Reply:
@peternied Read the conversation from here and below: https://github.com/opensearch-project/security/issues/2263#issuecomment-1603344501

This gets a copy of the set so that its not susceptible to concurrent modification. The better approach is to set the roles once and remove `addSecurityRoles` if its possible.

Reply:
> This gets a copy of the set so that its not susceptible to concurrent modification.

That is not correct, `user.getSecurityRoles().stream().collect(...)` still iterates over each item to make the copy.

Reply:
In to strict sense, while this code is running, the iterator is susceptible to concurrent modification exception.

If we look at the Sets.union(...) [[source code](https://github.com/google/guava/blob/f2f4f864d720ad03a045c9b918fe06cb11a65832/guava/src/com/google/common/collect/Sets.java#L713)], it is creating a view over both collections, so this change would use the resulting in memory 'ImmutableSet' and lower the likelihood for this exception.

Sets.union(...) earns a 👎 since on interaction its doing duplicate filtering ever time and doesn't make it obvious it is creating a side effect.